### PR TITLE
Add integration with cartographer in ROS 2

### PR DESCRIPTION
**Pull request by pedro-fuoco**
This PR is dedicated to adding cartographer-ros integration and transforming this repository into a ROS 2 package. Following the instructions on the new README, the results are as follows:

Screenshot from 2023-08-04 17-44-38

### CODE REVIEW

**Comment by pedro-fuoco** [comment]
Here is a video of it working correctly (tested using waypoints provided by GCS)
cartographer

**Comment by pedro-fuoco** [comment]
Note that in the current state the cartographer odometry output isn't being fed into Ardupilot, this will be done in another PR

File: README.md [hunk 1]
```diff
@@ -1,73 +1,40 @@
...
+        * Set the ENABLE_DDS ardupilot parameter manually if it isn't set
 
-This package contains a main launch file launching all the required launch files in a go.
+## Instalation
 
-Simply,
+Clone this repository into your ros2 workspace alongside ardupilot_gz
+```bash
+cd ~/ros2_ws/src
+git clone git@github.com:pedro-fuoco/ardupilot_ros.git -b ros2-package
```
Comment (Ryanf55) [inline README.md]: Update this to `ardupilot/ardupilot_ros -b humble`?

File: README.md [hunk 2]
```diff
@@ -1,73 +1,40 @@
...
+```
+Build it with colcon build
```
Comment (Ryanf55) [inline README.md:20]: `Build ardupilot_ros and ardupilot_gz with colcon build`? Unclear why you need --packages-select later

File: README.md [hunk 3]
```diff
@@ -1,73 +1,40 @@
...
 
-### On 1st Terminal
-````
-roslaunch ardupilot_ros gzbo.launch #for quadrotor
-````
-or
-````
-roslaunch ardupilot_ros gzbo_rover.launch #for rover
-````
+This simulation has an iris copter equipped with a 360 degrees 2d lidar in a maze world, to launch it run:
```
Comment (Ryanf55) [inline README.md]: capitalize `2d` to `2D`

File: launch/cartographer.launch.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        gz_sim_resource_path = os.environ["GZ_SIM_RESOURCE_PATH"]
+
+        if "SDF_PATH" in os.environ:
+            sdf_path = os.environ["SDF_PATH"]
+            os.environ["SDF_PATH"] = sdf_path + ":" + gz_sim_resource_path
+        else:
+            os.environ["SDF_PATH"] = gz_sim_resource_path
+
+    # Load SDF file.
+    sdf_file = os.path.join(
```
Comment (Ryanf55) [inline launch/cartographer.launch.py]: Can we switch to pathlib?

Comment (Ryanf55) [inline launch/cartographer.launch.py]: Can we use pathlib?

File: package.xml [hunk 5]
```diff
@@ -1,72 +1,18 @@
...
-  <!-- <maintainer email="jane.doe@example.com">Jane Doe</maintainer> -->
-  <maintainer email="snkt@todo.todo">snkt</maintainer>
-
-
-  <!-- One license tag required, multiple allowed, one license per tag -->
-  <!-- Commonly used license strings: -->
-  <!--   BSD, MIT, Boost Software License, GPLv2, GPLv3, LGPLv2.1, LGPLv3 -->
+  <description>The ardupilot ROS 2 use cases package</description>
+  <maintainer email="pedrofuoco6@gmail.com">Pedro Fuoco</maintainer>
   <license>TODO</license>
```
Comment (Ryanf55) [inline package.xml]: GPLv3?

File: setup.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        (os.path.join('share', package_name, 'launch'), glob(os.path.join('launch', '*launch.[pxy][yma]*'))),
+        (os.path.join('share', package_name, 'config'), glob(os.path.join('config', '*.lua'))),
+        (os.path.join('share', package_name, 'rviz'), glob(os.path.join('rviz', '*.rviz'))),
+    ],
+    install_requires=['setuptools'],
+    zip_safe=True,
+    maintainer='Pedro Fuoco',
+    maintainer_email='pedrofuoco6@gmail.com',
+    description='The ardupilot ROS 2 use cases package',
+    license='TODO',
```
Comment (Ryanf55) [inline setup.py]: Add license

**Review by Ryanf55 — COMMENTED** [review]
Looks great so far. I'll report back with testing results when I get some more time.

File: README.md (hunk 2 above)
Comment (pedro-fuoco) [reply README.md:20]: I assume `ardupilot_gz` is already built since it is listed as a requirement

File: package.xml (hunk 5 above)
Comment (pedro-fuoco) [reply package.xml]: I'm not sure what happened to the license, maybe it got lost in a sync at some point ? sorry about that

File: launch/cartographer.launch.py (hunk 4 above)
Comment (pedro-fuoco) [reply launch/cartographer.launch.py]: Will do! 
Did a little research as to why `pathlib` is preferred. I will leave this link here for future reference: https://stackoverflow.com/questions/67112343/pathlib-vs-os-path-join-in-python

File: README.md [hunk 7]
```diff
@@ -1,73 +1,40 @@
...
+* [Gazebo Garden](https://gazebosim.org/docs/garden/install)
 
-## Usage 2 :
+* [ardupilot_gz](https://github.com/ArduPilot/ardupilot_gz)
+    * Here are two workarounds for know issues in the installation of `ardupilot_gz` (they should be fixed soon)
+        * [Handle microxrceddsgen dependency](https://github.com/ArduPilot/ardupilot_gz/issues/19)
+        * Set the ENABLE_DDS ardupilot parameter manually if it isn't set
 
-This package contains a main launch file launching all the required launch files in a go.
+## Instalation
```
Comment (Ryanf55) [inline README.md]: Missing L

File: launch/cartographer.launch.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    use_sim_time_arg = DeclareLaunchArgument("use_sim_time", default_value="true")
+    rviz_arg = DeclareLaunchArgument(
+        "rviz", default_value="true", description="Open RViz."
+    )
+    # Robot description.
+
+    # Ensure `SDF_PATH` is populated as `sdformat_urdf`` uses this rather
+    # than `GZ_SIM_RESOURCE_PATH` to locate resources.
+    if "GZ_SIM_RESOURCE_PATH" in os.environ:
+        gz_sim_resource_path = os.environ["GZ_SIM_RESOURCE_PATH"]
```
Comment (Ryanf55) [inline launch/cartographer.launch.py]: Can simplify to this, and remove the if statement. 
gz_sim_resource_path = os.environ.get(["GZ_SIM_RESOURCE_PATH"])

File: launch/cartographer.launch.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+        if "SDF_PATH" in os.environ:
+            sdf_path = os.environ["SDF_PATH"]
+            os.environ["SDF_PATH"] = sdf_path + ":" + gz_sim_resource_path
```
Comment (Ryanf55) [inline launch/cartographer.launch.py]: Try `":".join([sdf_path, gz_sim_resource_path])`

File: launch/cartographer.launch.py (hunk 4 above)
Comment (Ryanf55) [reply launch/cartographer.launch.py]: Did you forget to push the change?

File: config/cartographer.lua [hunk 10]
```diff
@@ -52,4 +52,4 @@ POSE_GRAPH.constraint_builder.global_localization_min_score = 0.65
 POSE_GRAPH.optimization_problem.huber_scale = 1e2
 POSE_GRAPH.optimize_every_n_nodes = 30
 
-return options
+return options
```
Comment (Ryanf55) [inline config/cartographer.lua]: Missing newline. Please rebase on top of the main development branch and then `pre-commit run all-files` which will pick up any files in this PR.

File: launch/cartographer.launch.py (hunk 4 above)
Comment (pedro-fuoco) [reply launch/cartographer.launch.py]: Had an error, just fixed it

File: launch/cartographer.launch.py (hunk 8 above)
Comment (pedro-fuoco) [reply launch/cartographer.launch.py]: Completely removed this section, as robot_state_publisher was moved to `ardupilot_gz`

File: launch/cartographer.launch.py (hunk 9 above)
Comment (pedro-fuoco) [reply launch/cartographer.launch.py]: Completely removed this section, as robot_state_publisher was moved to `ardupilot_gz`

**Comment by pedro-fuoco** [comment]
Squashed and edited the commits, this is how the final version looks
Screenshot from 2023-08-11 12-12-53

Thanks a lot for the help @Ryanf55 !

**Review by Ryanf55 — APPROVED** [review]

**Pull request merged**